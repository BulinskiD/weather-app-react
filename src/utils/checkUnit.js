const checkUnit = (unit, setUnit) => {
    const storage = window.localStorage;
    let unitParam = storage.getItem("unit");
    /**** If unit param is set in localStorage, set it in state, else use default value ****/
    if(unitParam)
        setUnit(unitParam);
    else
        unitParam = unit;

    return unitParam;
}

export default checkUnit;
