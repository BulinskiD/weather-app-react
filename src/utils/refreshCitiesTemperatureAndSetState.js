//Refresh cities
import forecastApi from "../api/forecastApi";
import calculateAvg from "./calculateAvg";

export default (units, cities, setCities) => {
    const citiesPromises = cities.map(async ({id}) => {
        try {
            return await forecastApi.get('', {params: {id, units }});
        } catch(error) {
            /** TODO Handle errors here **/
        }
    });
    Promise.all(citiesPromises).then(newCities => {
        const newState = newCities.map(cityObj => {
            console.log("cityObj", cityObj);
            return {
                id: cityObj.data.city.id,
                name: cityObj.data.city.name,
                temperature: calculateAvg(cityObj.data.list)
            }
        })
        setCities(newState);
    }).catch(error => {/** TODO Handle error here **/});
}
