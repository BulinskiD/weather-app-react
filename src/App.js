import 'bootstrap/dist/css/bootstrap.css';

import React, {useEffect} from 'react';
import ReactDOM from 'react-dom';
import {BrowserRouter, Route} from "react-router-dom";
import Container from 'react-bootstrap/Container';

import Header from './header'
import MainPage from './main-page/index';
import Settings from './settings';
import Details from './details';
import UnitContext from './context/unit-context';
import LoadingContext from './context/loading-context';
import ErrorModal from "./shared/error-modal";
import useCities from "./hooks/city-hook";
import useUnit from "./hooks/unit-hook";
import checkUnit from "./utils/checkUnit";
import refreshCitiesTemperatureAndSetState from "./utils/refreshCitiesTemperatureAndSetState";

export default () => {

    const [loading, setLoading, onAddCity, onRemoveCity, cities, setCities, error, setError] = useCities();
    const [unit, toggleUnit, setUnit] = useUnit();

    /**** On load ****/
    useEffect(() => {
            /**** Get cities and unit from localStorage ****/
            const citiesFromStorage = JSON.parse(localStorage.getItem("cities"));
            const unitParam = checkUnit(unit, setUnit);

            /**** If online refresh temperatures on page load else if offline set cities from localStorage if available ****/
            if(citiesFromStorage && navigator.onLine)
                refreshCitiesTemperatureAndSetState(unitParam, citiesFromStorage, setCities, setError, setLoading);
            else if (citiesFromStorage && !navigator.onLine) {
                setCities(citiesFromStorage);
                setLoading(false);
            }
            else {
                setLoading(false);
            }
        },
        // eslint-disable-next-line
        []);

    /**** Refresh cities on unit change ****/
    useEffect( ()=>{
            if(cities.length !== 0) {
                refreshCitiesTemperatureAndSetState(unit, cities, setCities, setError, setLoading);
            }
        }, // eslint-disable-next-line
        [unit]);


    /**** Render content ****/
    return (
            <BrowserRouter>
                <LoadingContext.Provider value={{loading}}>
                    <UnitContext.Provider value={{unit, toggleUnit}}>
                        <Header />
                        <Container>
                            <Route path="/" exact component={() => <MainPage onRemoveCity={onRemoveCity} unit={unit} onAddCity={onAddCity} cities={cities} />}></Route>
                            <Route path="/settings" component={Settings} />
                            <Route path="/details/:id" component={(props) => <Details {...props} unit={unit} />} />
                        </Container>
                        {ReactDOM.createPortal(<ErrorModal show={error ? true : false} onClose={() => setError(null)}>{error}</ErrorModal>,
                            document.getElementById("root"))}
                    </UnitContext.Provider>
                </LoadingContext.Provider>
            </BrowserRouter>
    );
}

