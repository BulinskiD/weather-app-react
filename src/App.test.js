import React from 'react';
import { act } from 'react-dom/test-utils';
import refreshCitiesTemperatureAndSetState from './utils/refreshCitiesTemperatureAndSetState';
import checkUnit from './utils/checkUnit';
import useCities from "./hooks/city-hook";
import useUnit from "./hooks/unit-hook";
import App from './App';
import {shallow, mount} from 'enzyme';

jest.mock('./hooks/city-hook');
jest.mock('./hooks/unit-hook');
jest.mock('./utils/refreshCitiesTemperatureAndSetState');
jest.mock('./utils/checkUnit');

const loading = true;
const setLoading = jest.fn();
const onAddCity = jest.fn();
const onRemoveCity = jest.fn();
let cities = [];
const setCities = jest.fn();
const error = null;
const setError = jest.fn();

useCities.mockReturnValue([loading, setLoading, onAddCity, onRemoveCity, cities, setCities, error, setError]);

let unit = 'metric';
const toggleUnit = jest.fn();
const setUnit = jest.fn();

useUnit.mockReturnValue([unit, toggleUnit, setUnit]);

const navigatorMock = jest.spyOn(navigator, 'onLine', 'get');

const e = {preventDefault: ()=>{}};

// add a div with #root id for react.createPortal
const modalRoot = global.document.createElement('div');
modalRoot.setAttribute('id', 'root');
const body = global.document.querySelector('body');
body.appendChild(modalRoot);

beforeEach(() => {
    setLoading.mockClear();
    onAddCity.mockClear();
    onRemoveCity.mockClear();
    setCities.mockClear();
    setError.mockClear();
    toggleUnit.mockClear();
    setUnit.mockClear();
    checkUnit.mockClear();
});


describe('App', ()=>{

    it('should match snapshot', () => {
        const component = shallow(<App/>);
        expect(component).toMatchSnapshot();
    });

    it('should not call refreshCitiesTemperatureAndSetState when navigator is offline', ()=>{
        navigatorMock.mockReturnValue(false);
        localStorage.getItem.mockReturnValue(JSON.stringify([{ id: 1, name: "Gorlice", temperature: "10.00" }]));

        act(()=>{
            mount(<App />);
        });

        expect(refreshCitiesTemperatureAndSetState).toBeCalledTimes(0);
        expect(setCities).toBeCalledWith([{ id: 1, name: "Gorlice", temperature: "10.00" }]);
        expect(setLoading).toBeCalledTimes(1);
    });

    it('should call refreshCitiesTemperatureAndSetState when navigator is online', ()=>{
        navigatorMock.mockReturnValue(true);
        checkUnit.mockReturnValue(unit);
        localStorage.getItem.mockReturnValue(JSON.stringify([{ id: 1, name: "Gorlice", temperature: "10.00" }]));

        act(()=>{
           mount(<App />);
        });

        expect(checkUnit).toBeCalledTimes(1);
        expect(refreshCitiesTemperatureAndSetState).toBeCalledWith(unit, [{ id: 1, name: "Gorlice", temperature: "10.00" }], setCities, setError, setLoading);
        expect(setLoading).toBeCalledTimes(0);
    });

    it('should call only setLoading when there is no data in localStorage', ()=>{
        navigatorMock.mockReturnValue(true);
        checkUnit.mockReturnValue(unit);
        localStorage.getItem.mockReturnValue(null);

        act(()=>{
            mount(<App />);
        });

        expect(checkUnit).toBeCalledTimes(1);
        expect(setLoading).toBeCalledWith(false);
    });
});
