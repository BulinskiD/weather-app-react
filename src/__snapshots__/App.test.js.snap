// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<BrowserRouter>
  <ContextProvider
    value={
      Object {
        "loading": true,
      }
    }
  >
    <ContextProvider
      value={
        Object {
          "toggleUnit": [MockFunction],
          "unit": "metric",
        }
      }
    >
      <_default />
      <Container
        as="div"
        fluid={false}
      >
        <Route
          component={[Function]}
          exact={true}
          path="/"
        />
        <Route
          component={[Function]}
          path="/settings"
        />
        <Route
          component={[Function]}
          path="/details/:id"
        />
      </Container>
      <Portal
        containerInfo={
          <div
            id="root"
          />
        }
      >
        <ErrorModal
          onClose={[Function]}
          show={false}
        />
      </Portal>
    </ContextProvider>
  </ContextProvider>
</BrowserRouter>
`;
